File: lotto.js
(function () {
  const MAX_GAMES = 5;
  const SELECTION_LABELS = {
    q: "auto",
    m: "manual"
  };

  function ok(data) {
    return { ok: true, data };
  }

  function error(message, extra = {}) {
    return { ok: false, error: message, ...extra };
  }

  function selectionLabel(selectionCode) {
    return SELECTION_LABELS[selectionCode];
  }

  function extractVValue(raw) {
    if (raw == null) {
      return "";
    }

    const text = String(raw).trim();
    if (!text) {
      return "";
    }

    try {
      const url = new URL(text, window.location.href);
      const value = url.searchParams.get("v");
      if (value) {
        return decodeURIComponent(value).trim();
      }
    } catch (event) {
      // Fall through to query-string parsing below.
    }

    const queryText = text.startsWith("?") ? text.slice(1) : text;
    if (queryText.startsWith("v=") || queryText.includes("&v=")) {
      const params = new URLSearchParams(queryText);
      const value = params.get("v");
      if (value) {
        return decodeURIComponent(value).trim();
      }
    }

    const match = text.match(/[?&]v=([^&#]+)/);
    if (match) {
      return decodeURIComponent(match[1]).trim();
    }

    return text;
  }

  function validateGame(numbers, gameId) {
    if (numbers.length !== 6) {
      return `게임 ${gameId}의 번호는 6개여야 합니다.`;
    }

    if (new Set(numbers).size !== 6) {
      return `게임 ${gameId}에 중복 번호가 있습니다.`;
    }

    const invalid = numbers.find((number) => number < 1 || number > 45);
    if (invalid != null) {
      return `게임 ${gameId}의 번호 ${String(invalid).padStart(2, "0")}는 1~45 범위를 벗어났습니다.`;
    }

    return null;
  }

  function parseGameSegment(selectionCode, segment, gameId) {
    const label = selectionLabel(selectionCode);
    if (!label) {
      return error(`게임 ${gameId}의 선택 유형을 알 수 없습니다.`);
    }

    if (!/^\d{12}$/.test(segment || "")) {
      return error(`게임 ${gameId}의 번호 데이터가 올바르지 않습니다.`);
    }

    const numbers = [];
    for (let index = 0; index < 12; index += 2) {
      numbers.push(Number(segment.slice(index, index + 2)));
    }

    const validationError = validateGame(numbers, gameId);
    if (validationError) {
      return error(validationError);
    }

    return ok({
      id: gameId,
      selectionType: label,
      selectionCode,
      numbers
    });
  }

  function parseGames(value) {
    if (value.length < 17) {
      return error("QR 번호 데이터가 부족합니다.");
    }

    const roundText = value.slice(0, 4);
    if (!/^\d{4}$/.test(roundText)) {
      return error("회차 정보가 올바르지 않습니다.");
    }

    const body = value.slice(4);
    if (!body) {
      return error("게임 번호 데이터가 없습니다.");
    }

    const games = [];
    let cursor = 0;

    while (cursor < body.length && games.length < MAX_GAMES) {
      const selectionCode = body.slice(cursor, cursor + 1);

      if (!SELECTION_LABELS[selectionCode]) {
        if (games.length > 0) {
          break;
        }
        return error(`지원하지 않는 선택 유형 '${selectionCode}'가 포함되어 있습니다.`);
      }

      if (cursor + 13 > body.length) {
        const gameId = String.fromCharCode("A".charCodeAt(0) + games.length);
        return error(`게임 ${gameId}의 번호 데이터가 올바르지 않습니다.`);
      }

      const segment = body.slice(cursor + 1, cursor + 13);
      const gameId = String.fromCharCode("A".charCodeAt(0) + games.length);
      const parsed = parseGameSegment(selectionCode, segment, gameId);
      if (!parsed.ok) {
        return parsed;
      }

      games.push(parsed.data);
      cursor += 13;
    }

    if (games.length === 0) {
      return error("게임 번호 데이터가 없습니다.");
    }

    const extraData = body.slice(cursor);

    return ok({
      round: Number(roundText),
      games,
      extraData,
      hasExtraData: Boolean(extraData)
    });
  }

  function parseLottoQr(raw) {
    const value = extractVValue(raw);
    if (!value) {
      return error("QR URL 또는 v 값을 입력해 주세요.");
    }

    const parsed = parseGames(value);
    if (!parsed.ok) {
      return parsed;
    }

    parsed.data.raw = String(raw || "").trim();
    parsed.data.value = value;
    return parsed;
  }

  function collectExcludedNumbers(payload = {}) {
    const parsed = payload.parsed || {};
    const games = parsed.games || [];
    const excluded = new Set();

    if (payload.excludeAllScanned) {
      games.forEach((game) => {
        (game.numbers || []).forEach((number) => excluded.add(number));
      });
    }

    const excludedGameIds = new Set(payload.excludedGameIds || []);
    games.forEach((game) => {
      if (excludedGameIds.has(game.id)) {
        (game.numbers || []).forEach((number) => excluded.add(number));
      }
    });

    (payload.manualExcludedNumbers || []).forEach((number) => {
      const value = Number(number);
      if (Number.isInteger(value) && value >= 1 && value <= 45) {
        excluded.add(value);
      }
    });

    return Array.from(excluded).sort((a, b) => a - b);
  }

  function generateNumbers(payload = {}) {
    const excludedNumbers = collectExcludedNumbers(payload);
    const excluded = new Set(excludedNumbers);
    const candidates = [];

    for (let number = 1; number <= 45; number += 1) {
      if (!excluded.has(number)) {
        candidates.push(number);
      }
    }

    if (candidates.length < 6) {
      return error("제외 조건이 너무 많아 번호를 생성할 수 없습니다.", {
        excludedNumbers,
        availableCount: candidates.length
      });
    }

    const pool = [...candidates];
    const numbers = [];
    for (let index = 0; index < 6; index += 1) {
      const selectedIndex = Math.floor(Math.random() * pool.length);
      numbers.push(pool[selectedIndex]);
      pool.splice(selectedIndex, 1);
    }

    numbers.sort((a, b) => a - b);

    return {
      ok: true,
      numbers,
      excludedNumbers,
      availableCount: candidates.length
    };
  }

  window.Lotto = {
    parseLottoQr,
    generateNumbers,
    collectExcludedNumbers
  };
})();

File: lotto.py
import json
import random
import re
from urllib.parse import parse_qs, unquote, urlparse


SELECTION_LABELS = {
    "q": "auto",
    "m": "manual",
}


def _ok(data):
    return {"ok": True, "data": data}


def _error(message, **extra):
    result = {"ok": False, "error": message}
    result.update(extra)
    return result


def selection_label(selection_code):
    return SELECTION_LABELS.get(selection_code)


def extract_v_value(raw):
    if raw is None:
        return ""

    text = str(raw).strip()
    if not text:
        return ""

    parsed = urlparse(text)
    query = parsed.query

    if query:
        values = parse_qs(query).get("v")
        if values:
            return unquote(values[0]).strip()

    if text.startswith("?"):
        values = parse_qs(text[1:]).get("v")
        if values:
            return unquote(values[0]).strip()

    if text.startswith("v="):
        values = parse_qs(text).get("v")
        if values:
            return unquote(values[0]).strip()

    match = re.search(r"(?:[?&]v=)([^&#]+)", text)
    if match:
        return unquote(match.group(1)).strip()

    return text


def validate_game(numbers, game_id):
    if len(numbers) != 6:
        return "게임 %s의 번호는 6개여야 합니다." % game_id

    if len(set(numbers)) != 6:
        return "게임 %s에 중복 번호가 있습니다." % game_id

    for number in numbers:
        if number < 1 or number > 45:
            return "게임 %s의 번호 %02d는 1~45 범위를 벗어났습니다." % (game_id, number)

    return None


def parse_game_segment(selection_code, segment, game_id):
    label = selection_label(selection_code)
    if label is None:
        return _error("게임 %s의 선택 유형을 알 수 없습니다." % game_id)

    if not re.fullmatch(r"\d{12}", segment or ""):
        return _error("게임 %s의 번호 데이터가 올바르지 않습니다." % game_id)

    numbers = [int(segment[index:index + 2]) for index in range(0, 12, 2)]
    validation_error = validate_game(numbers, game_id)
    if validation_error:
        return _error(validation_error)

    return _ok({
        "id": game_id,
        "selectionType": label,
        "selectionCode": selection_code,
        "numbers": numbers,
    })


def parse_games(value):
    if len(value) < 17:
        return _error("QR 번호 데이터가 부족합니다.")

    round_text = value[:4]
    if not re.fullmatch(r"\d{4}", round_text):
        return _error("회차 정보가 올바르지 않습니다.")

    body = value[4:]
    if not body:
        return _error("게임 번호 데이터가 없습니다.")

    games = []
    cursor = 0
    game_index = 0

    while cursor < len(body):
        selection_code = body[cursor:cursor + 1]
        if selection_code not in SELECTION_LABELS:
            return _error("지원하지 않는 선택 유형 '%s'가 포함되어 있습니다." % selection_code)

        segment = body[cursor + 1:cursor + 13]
        game_id = chr(ord("A") + game_index)
        parsed = parse_game_segment(selection_code, segment, game_id)
        if not parsed["ok"]:
            return parsed

        games.append(parsed["data"])
        cursor += 13
        game_index += 1

    if not games:
        return _error("게임 번호 데이터가 없습니다.")

    return _ok({
        "round": int(round_text),
        "games": games,
    })


def parse_lotto_qr(raw):
    value = extract_v_value(raw)
    if not value:
        return _error("QR URL 또는 v 값을 입력해 주세요.")

    parsed = parse_games(value)
    if not parsed["ok"]:
        return parsed

    parsed["data"]["raw"] = str(raw).strip()
    parsed["data"]["value"] = value
    return parsed


def collect_excluded_numbers(payload):
    parsed = payload.get("parsed") or {}
    games = parsed.get("games") or []
    excluded = set()

    if payload.get("excludeAllScanned"):
        for game in games:
            excluded.update(game.get("numbers") or [])

    excluded_game_ids = set(payload.get("excludedGameIds") or [])
    for game in games:
        if game.get("id") in excluded_game_ids:
            excluded.update(game.get("numbers") or [])

    for number in payload.get("manualExcludedNumbers") or []:
        try:
            value = int(number)
        except (TypeError, ValueError):
            continue
        if 1 <= value <= 45:
            excluded.add(value)

    return sorted(excluded)


def generate_numbers(payload):
    excluded_numbers = collect_excluded_numbers(payload or {})
    excluded = set(excluded_numbers)
    candidates = [number for number in range(1, 46) if number not in excluded]

    if len(candidates) < 6:
        return _error(
            "제외 조건이 너무 많아 번호를 생성할 수 없습니다.",
            excludedNumbers=excluded_numbers,
            availableCount=len(candidates),
        )

    numbers = sorted(random.sample(candidates, 6))
    return {
        "ok": True,
        "numbers": numbers,
        "excludedNumbers": excluded_numbers,
        "availableCount": len(candidates),
    }


def parse_lotto_qr_json(raw):
    return json.dumps(parse_lotto_qr(raw), ensure_ascii=False)


def generate_numbers_json(payload_json):
    try:
        payload = json.loads(payload_json or "{}")
    except ValueError:
        return json.dumps(_error("생성 조건 데이터가 올바르지 않습니다."), ensure_ascii=False)

    return json.dumps(generate_numbers(payload), ensure_ascii=False)
